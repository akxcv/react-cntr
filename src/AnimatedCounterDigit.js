import React from 'react'
import ReactDOM from 'react-dom'
import AbstractCounterDigit from './AbstractCounterDigit'
const { number, string } = React.PropTypes

/**
 * Forces reflow of a given element.
 * @param {Element} element
 */
function forceReflow (element) {
  element.offsetHeight // eslint-disable-line no-unused-expressions
}

/**
 * Animated digit component.
 * Used when counter has an easing function.
 * @example
 * <AnimatedCounterDigit
 *   digit='5'
 *   radix={10}
 *   direction='down'
 *   digitMap={{}}
 *   digitWrapper={(digit) => digit}
 *   maxValue={9}
 *   easingFunction='ease-in'
 *   easingDuration={200}
 * />
 */
class AnimatedCounterDigit extends AbstractCounterDigit {
  /**
   * @property {string} direction - counting direction
   * @property {number} maxValue - maximum value used to build a digit lane
   * @property {string} easingFunction - easing function for transitions
   * @property {number} easingDuration - duration for digit transitions in milliseconds
   */
  static propTypes = {
    ...AbstractCounterDigit.propTypes,
    direction: string.isRequired,
    maxValue: number.isRequired,
    easingFunction: string.isRequired,
    easingDuration: number.isRequired
  }

  constructor (props) {
    super(props)

    // e.x.: 0..9 and another 0 => 9+2 = 11 digits total (see reset method)
    const singleDigitHeight = 100 / (this.props.maxValue + 2)
    const upperZeroPosition = -singleDigitHeight * (this.props.radix - this.props.maxValue - 1)
    const lowerZeroPosition = -singleDigitHeight * this.props.radix
    const initialZeroPosition = (this.props.direction === 'down')
      ? lowerZeroPosition : upperZeroPosition
    const finalZeroPosition = (this.props.direction === 'down')
      ? upperZeroPosition : lowerZeroPosition

    this.state = {
      initialZeroPosition,
      finalZeroPosition,
      singleDigitHeight,
      digitLane: this.buildDigitLane()
    }
  }

  componentDidMount () {
    this.reset({ target: ReactDOM.findDOMNode(this) })
  }

  /**
   * Changes vertical position of digit lane without triggering a transition.
   * In a lane like [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0] goes from first zero to last zero instantly.
   * @param {SyntheticEvent} event
   */
  reset = (event) => {
    if (this.props.digit === '0') {
      event.target.style.transitionProperty = 'none'
      event.target.style.transform = `translateY(${this.state.initialZeroPosition}%)`
      forceReflow(event.target)
      event.target.style.transitionProperty = ''
    }
  }

  /**
   * Creates a div for a digit lane member.
   * @param {string} digit
   * @param {number} key - index to use for React's key property
   * @return {ReactElement}
   */
  buildDigitDiv (digit, key) {
    return (
      <div key={key || digit} className='rollex-digit-lane-label' style={{ height: this.props.height }}>
        {this.decorateDigit(digit)}
      </div>
    )
  }

  /**
   * Builds the digit lane.
   * Digit lane contains all available digits.
   * It changes its vertical position in order to emulate counter rolling.
   * @return {ReactElement[]}
   */
  buildDigitLane () {
    var digitLane = []
    for (let i = 0; i <= this.props.maxValue; i++) {
      digitLane.push(this.buildDigitDiv(i))
    }
    digitLane.push(this.buildDigitDiv(0, this.props.radix + 1))
    return digitLane
  }

  /**
   * Calculates Y position for the digit.
   * @return {number} yPosition
   */
  getDigitPositionY () {
    const digitIndex = parseInt(this.props.digit, this.props.radix)
    if (digitIndex === 0) {
      return this.state.finalZeroPosition
    } else {
      return -this.state.singleDigitHeight * (digitIndex + (this.props.radix - this.props.maxValue - 1))
    }
  }

  /**
   * Renders the digit.
   * @return {ReactElement} digit
   */
  render () {
    const transform = `translateY(${this.getDigitPositionY()}%)`
    const style = {
      transform: transform,
      WebkitTransform: transform,
      MsTransform: transform,
      transitionTimingFunction: this.props.easingFunction,
      transitionDuration: `${this.props.easingDuration}ms`,
      transitionProperty: 'transform',
      willChange: 'transform'
    }

    return (
      <div className='rollex-digit' onTransitionEnd={this.reset} style={style}>
        {this.state.digitLane}
      </div>
    )
  }
}

export default AnimatedCounterDigit
