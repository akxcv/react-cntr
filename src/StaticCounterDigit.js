import React from 'react'
import AbstractCounterDigit from './AbstractCounterDigit'

/**
 * static digit component
 * used when no easing function is set
 * @example
 * <StaticCounterDigit
 *   digit='5'
 *   height={18}
 *   radix={10}
 *   digitMap={{}}
 *   digitWrapper={(digit) => digit}
 * />
 */
class StaticCounterDigit extends AbstractCounterDigit {
  /**
   * render
   * @return {ReactElement} digit
   */
  render () {
    return (
      <div className='rollex-digit'>
        {this.decorateDigit(this.props.digit)}
      </div>
    )
  }
}

export default StaticCounterDigit
