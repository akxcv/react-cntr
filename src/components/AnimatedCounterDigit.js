import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { digitPropTypes, decorateDigit } from '../helpers/counterDigitHelper'

/**
 * Forces reflow of a given element.
 * @param {Element} element
 */
function forceReflow (element) {
  element.offsetHeight // eslint-disable-line no-unused-expressions
}

/**
 * Animated digit component.
 * Used when counter has an easing function.
 * @example
 * <AnimatedCounterDigit
 *   digit='5'
 *   radix={10}
 *   direction='down'
 *   digitMap={{}}
 *   digitWrapper={(digit) => digit}
 *   maxValue={9}
 *   easingFunction='ease-in'
 *   easingDuration={200}
 * />
 */
class AnimatedCounterDigit extends React.Component {
  /**
   * @property {string} digit - digit to display
   * @property {number} radix
   * @property {Object} digitMap - a map for transforming particular digits
   * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
   * @property {string} direction - counting direction
   * @property {number} maxValue - maximum value used to build a digit lane
   * @property {string} easingFunction - easing function for transitions
   * @property {number} easingDuration - duration for digit transitions in milliseconds
   */
  static propTypes = {
    ...digitPropTypes,
    direction: PropTypes.string.isRequired,
    maxValue: PropTypes.number.isRequired,
    easingFunction: PropTypes.string.isRequired,
    easingDuration: PropTypes.number.isRequired
  }

  constructor (props) {
    super(props)

    // e.x.: 0..9 and another 0 => 9+2 = 11 digits total (see reset method)
    const singleDigitHeight = 100 / (props.maxValue + 2)
    const upperZeroPosition =
      -singleDigitHeight * (props.radix - props.maxValue - 1)
    const lowerZeroPosition = -singleDigitHeight * props.radix
    const initialZeroPosition = props.direction === 'down'
      ? lowerZeroPosition
      : upperZeroPosition
    const finalZeroPosition = props.direction === 'down'
      ? upperZeroPosition
      : lowerZeroPosition

    this.state = {
      initialZeroPosition,
      finalZeroPosition,
      singleDigitHeight,
      digitLane: this.buildDigitLane()
    }
  }

  componentDidMount () {
    this.reset({ target: ReactDOM.findDOMNode(this) })
  }

  /**
   * Changes vertical position of digit lane without triggering a transition.
   * In a lane like [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0] goes from first zero to last zero instantly.
   * @param {SyntheticEvent} event
   */
  reset = event => {
    if (this.props.digit === '0') {
      event.target.style.transitionProperty = 'none'
      event.target.style.transform = `translate3D(0, ${this.state.initialZeroPosition}%, 0)`
      forceReflow(event.target)
      event.target.style.transitionProperty = ''
    }
  }

  /**
   * Creates a div for a digit lane member.
   * @param {string} digit
   * @param {number} key - index to use for React's key property
   * @return {ReactElement}
   */
  buildDigitDiv (digit, key) {
    return (
      <div key={key || digit} className='rollex-digit-lane-label'>
        {decorateDigit(digit, this.props)}
      </div>
    )
  }

  /**
   * Builds the digit lane.
   * Digit lane contains all available digits.
   * It changes its vertical position in order to emulate counter rolling.
   * @return {ReactElement[]}
   */
  buildDigitLane () {
    var digitLane = []
    for (let i = 0; i <= this.props.maxValue; i++) {
      digitLane.push(this.buildDigitDiv(i))
    }
    digitLane.push(this.buildDigitDiv(0, this.props.radix + 1))
    return digitLane
  }

  /**
   * Calculates Y position for the digit.
   * @return {number} yPosition
   */
  getDigitPositionY () {
    const digitIndex = parseInt(this.props.digit, this.props.radix)
    if (digitIndex === 0) {
      return this.state.finalZeroPosition
    } else {
      return (
        -this.state.singleDigitHeight *
        (digitIndex + (this.props.radix - this.props.maxValue - 1))
      )
    }
  }

  /**
   * Renders the digit.
   * @return {ReactElement} digit
   */
  render () {
    const transform = `translate3D(0, ${this.getDigitPositionY()}%, 0)`
    const style = {
      transform: transform,
      WebkitTransform: transform,
      MsTransform: transform,
      transitionTimingFunction: this.props.easingFunction,
      transitionDuration: `${this.props.easingDuration}ms`,
      transitionProperty: 'transform',
      willChange: 'transform'
    }

    return (
      <div className='rollex-digit' onTransitionEnd={this.reset} style={style}>
        {this.state.digitLane}
      </div>
    )
  }
}

export default AnimatedCounterDigit
