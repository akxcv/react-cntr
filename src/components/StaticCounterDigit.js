import React from 'react'
import { digitPropTypes, decorateDigit } from '../helpers/counterDigitHelper'

/**
 * @property {string} digit - digit to display
 * @property {number} radix
 * @property {Object} digitMap - a map for transforming particular digits
 * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
 */
StaticCounterDigit.propTypes = digitPropTypes

/**
 * Static digit component.
 * Used when no easing function is set for a counter.
 * @example
 * <StaticCounterDigit
 *   digit='5'
 *   radix={10}
 *   digitMap={{}}
 *   digitWrapper={(digit) => digit}
 * />
 */
export default function StaticCounterDigit (props) {
  return (
    <div className='rollex-digit'>
      {decorateDigit(props.digit, props)}
    </div>
  )
}
