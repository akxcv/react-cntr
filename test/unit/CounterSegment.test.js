import { shallowToJson } from 'enzyme-to-json'
import { CounterSegmentBuilder as Builder } from './support/builders'
import AnimatedCounterDigit from '../../src/components/AnimatedCounterDigit'
import StaticCounterDigit from '../../src/components/StaticCounterDigit'

describe('snaphots', function () {
  var staticComponent, animatedComponent

  beforeAll(function () {
    staticComponent = Builder.shallow({ easingFunction: null })
    animatedComponent = Builder.shallow({ easingFunction: 'ease-in' })
  })

  it('matches snapshots', function () {
    const staticTree = shallowToJson(staticComponent)
    const animatedTree = shallowToJson(animatedComponent)
    expect(staticTree).toMatchSnapshot()
    expect(animatedTree).toMatchSnapshot()
  })
})

describe('rendering', function () {
  var staticComponent, animatedComponent

  beforeAll(function () {
    staticComponent = Builder.mount({ easingFunction: null })
    animatedComponent = Builder.mount({ easingFunction: 'ease-in' })
  })

  it('renders a CounterDigit for each digit', function () {
    expect(staticComponent.find(StaticCounterDigit).length).toBe(2)
    expect(staticComponent.find(AnimatedCounterDigit).length).toBe(0)
    expect(animatedComponent.find(AnimatedCounterDigit).length).toBe(2)
    expect(animatedComponent.find(StaticCounterDigit).length).toBe(0)
  })

  it('passes correct maxValue to digits', function () {
    const animatedDigits = animatedComponent.find(AnimatedCounterDigit)
    expect(animatedDigits.at(0).props().maxValue).toBe(5)
    expect(animatedDigits.at(1).props().maxValue).toBe(9)

    const dozenalComponent = Builder.mount({
      easingFunction: 'ease-in',
      radix: 12
    })
    const dozenalDigits = dozenalComponent.find(AnimatedCounterDigit)
    expect(dozenalDigits.at(0).props().maxValue).toBe(4)
    expect(dozenalDigits.at(1).props().maxValue).toBe(11)

    const daysComponent = Builder.mount({
      easingFunction: 'ease-in',
      period: 'days'
    })
    const daysComponentDigits = daysComponent.find(AnimatedCounterDigit)
    expect(daysComponentDigits.at(0).props().maxValue).toBe(9)
    expect(daysComponentDigits.at(1).props().maxValue).toBe(9)
  })
})
