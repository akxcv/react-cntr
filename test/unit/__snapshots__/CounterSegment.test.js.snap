// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snaphots matches snapshots 1`] = `
<div
  className="rollex-segment"
>
  <div
    className="rollex-digits"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <StaticCounterDigit
      digit="0"
      digitMap={Object {}}
      digitWrapper={[Function]}
      radix={10}
    />
    <StaticCounterDigit
      digit="0"
      digitMap={Object {}}
      digitWrapper={[Function]}
      radix={10}
    />
  </div>
  <div
    className="rollex-label"
  >
    seconds
  </div>
</div>
`;

exports[`snaphots matches snapshots 2`] = `
<div
  className="rollex-segment"
>
  <div
    className="rollex-digits"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <AnimatedCounterDigit
      digit="0"
      digitMap={Object {}}
      digitWrapper={[Function]}
      direction="down"
      easingDuration={300}
      easingFunction="ease-in"
      maxValue={5}
      radix={10}
    />
    <AnimatedCounterDigit
      digit="0"
      digitMap={Object {}}
      digitWrapper={[Function]}
      direction="down"
      easingDuration={300}
      easingFunction="ease-in"
      maxValue={9}
      radix={10}
    />
  </div>
  <div
    className="rollex-label"
  >
    seconds
  </div>
</div>
`;
